spring-messaging/src/main/java/org/springframework/messaging/converter/ProtobufMessageConverter.java

```diff
@@ -0,0 +1,368 @@
+	 */
+	private Message.Builder getMessageBuilder(Class<?> clazz) {
+		try {
+			assert supports(clazz);
+
+			Method method = methodCache.get(clazz);
+			if (method == null) {
+				method = clazz.getMethod("newBuilder");
+				methodCache.put(clazz, method);
+			}
+			return (Message.Builder) method.invoke(clazz);
+		} catch (Exception ex) {
+			throw new MessageConversionException(
+					"Invalid Protobuf Message type: no invocable newBuilder() method on " + clazz, ex);
+		}
+	}
+
+
+	@Override
+	protected boolean canConvertTo(Object payload, @Nullable MessageHeaders headers) {
+		MimeType mimeType = getMimeType(headers);
+		return (super.canConvertTo(payload, headers) ||
+				this.protobufFormatSupport != null && this.protobufFormatSupport.supportsWriteOnly(mimeType));
+	}
+
+	@Override
+	protected Object convertFromInternal(org.springframework.messaging.Message<?> message, Class<?> targetClass, @Nullable Object conversionHint) {
+		MimeType contentType = getMimeType(message.getHeaders());
+		final Object payload = message.getPayload();
+
+		if (contentType == null) {
+			contentType = PROTOBUF;
+		}
+
+		Charset charset = contentType.getCharset();
+		if (charset == null) {
+			charset = DEFAULT_CHARSET;
+		}
+
+		Message.Builder builder = getMessageBuilder(targetClass);
+
+		try {
+			if (PROTOBUF.isCompatibleWith(contentType)) {
+				builder.mergeFrom((byte[]) payload, this.extensionRegistry);
+			} else if (protobufFormatSupport != null) {
+				this.protobufFormatSupport.merge(
+						message, charset, contentType, this.extensionRegistry, builder);
+			}
+
+		} catch (IOException e) {
+			throw new MessageConversionException(message, "Could not read proto message" + e.getMessage(), e);
+		}
+
+		return builder.build();
+	}
+
+
+	@Override
+	protected Object convertToInternal(Object payload, @Nullable MessageHeaders headers, @Nullable Object conversionHint) {
+
+		final Message message = (Message) payload;
+
+		MimeType contentType = getMimeType(headers);
+		if (contentType == null) {
+			contentType = PROTOBUF;
+
+		}
+
+		Charset charset = contentType.getCharset();
+		if (charset == null) {
+			charset = DEFAULT_CHARSET;
+		}
+
+		try {
+			if (PROTOBUF.isCompatibleWith(contentType)) {
+				ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+				message.writeTo(byteArrayOutputStream);
+				payload = byteArrayOutputStream.toByteArray();
+			} else if (this.protobufFormatSupport != null) {
+				ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+				this.protobufFormatSupport.print(message, outputStream, contentType, charset);
+				payload = new String(outputStream.toByteArray(), charset);
+			}
+		} catch (IOException e) {
+			throw new MessageConversionException("Could not write proto message" + e.getMessage(), e);
+
+		}
+		return payload;
+	}
+
+	@Override
+	protected boolean supports(Class<?> clazz) {
+		return Message.class.isAssignableFrom(clazz);
+	}
```

Comment:
This method should be up by `canConvertTo` as it will be invoked from it.

Reply:
.Could you describe in more detail here? This function is called when I call `super.canConvertTo(payload, headers)` in method `canConvertTo`

Reply:
It's a minor comment about the order of methods. Logically this one should be right below `canConvert` but isn't. No big deal either way. I can polish that after you're done.

Reply:
thanks. got it